File: java-bidir/CallbackI.java
import Example.*;

class CallbackI extends _CallbackDisp implements java.lang.Runnable {
  private Ice.Communicator _broker;
  private boolean _destroy = false;
  private java.util.List<PrinterPrx> _clients =
    new java.util.ArrayList<PrinterPrx>();

  CallbackI(Ice.Communicator communicator) {
    _broker = communicator;
  }

  synchronized public void
  destroy() {
    System.out.println("destroying callback sender");
    _destroy = true;

    notify();

  }

  @Override
  public void
  attach(Ice.Identity ident, Ice.Current current) {
    System.out.println("new printer '" + _broker.identityToString(ident) + "'");

    Ice.ObjectPrx base = current.con.createProxy(ident);
    PrinterPrx client = PrinterPrxHelper.uncheckedCast(base);
    _clients.add(client);
  }

  @Override
  public void
  run() {
    int num = 0;

    while(true) {
      java.util.List<PrinterPrx> clients;

      synchronized(this) {
        try {
          wait(2000);
        }

        catch(java.lang.InterruptedException ex) {}

        if(_destroy)
          break;

        clients = new java.util.ArrayList<PrinterPrx>(_clients);
      }

      if(!clients.isEmpty()) {
        ++num;

        for(PrinterPrx remote_printer : clients) {
          try {
            remote_printer.write("text " + num);
          }

          catch(Ice.Exception ex) {
            System.out.println("removing client '" + _broker.identityToString(
                                 remote_printer.ice_getIdentity()) + "'");

            synchronized(this) {
              _clients.remove(remote_printer);
            }
          }
        }
      }

    }
  }
}

File: cpp-bidir/CallbackI.cpp
#include <string>
#include <iostream>
#include <Ice/Connection.h>
#include "CallbackI.h"


using namespace Example;
using namespace std;

CallbackI::CallbackI(const Ice::CommunicatorPtr& broker) :
  _broker(broker), _destroy(false) {}

void
CallbackI::destroy() {
  {
    IceUtil::Monitor<IceUtil::Mutex>::Lock lck(*this);

    cout << "destroying callback sender" << endl;
    _destroy = true;

    notify();
  }

  getThreadControl().join();
}


void
CallbackI::attach(const Ice::Identity& ident,
                              const Ice::Current& current) {

  IceUtil::Monitor<IceUtil::Mutex>::Lock lck(*this);


  cout << "new printer '" << _broker->identityToString(ident) << "'" <<
    endl;

  Example::PrinterPrx client = Example::PrinterPrx::uncheckedCast(
    current.con->createProxy(ident));
  _clients.insert(client);
}

void
CallbackI::run() {
  int num = 0;

  while(true) {
    set<PrinterPrx> clients;

    {
      IceUtil::Monitor<IceUtil::Mutex>::Lock lck(*this);
      timedWait(IceUtil::Time::seconds(2));

      if(_destroy)
        break;

      clients = _clients;
    }

    if(!clients.empty()) {
      ++num;

      for(set<PrinterPrx>::iterator p=clients.begin(); p!= clients.end(); ++p) {
        try {
          stringstream stream;
          stream << "text " << num;
          (*p)->write(stream.str());
        }

        catch(const Ice::Exception& ex) {
          cerr << "removing client '" << _broker->identityToString(
            (*p)->ice_getIdentity()) << "'" << endl;

          IceUtil::Monitor<IceUtil::Mutex>::Lock lck(*this);
          _clients.erase(*p);
        }
      }
    }
  }
}
